curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.english.md

```diff
@@ -24,13 +24,81 @@ Start this project on Glitch using <a href='https://glitch.com/edit/#!/remix/clo
 
 ```yml
 tests:
-  - text: I can pass a URL as a parameter and I will receive a shortened URL in the JSON response.
-    testString: ''
-  - text: 'If I pass an invalid URL that doesn''t follow the valid http://www.example.com format, the JSON response will contain an error instead.'
-    testString: ''
-  - text: 'When I visit that shortened URL, it will redirect me to my original link.'
-    testString: ''
+  - text: You should provide your own project, not the example URL.
+    testString: "getUserInput => {
+      const url = getUserInput('url');
+      assert(!(new RegExp('.*/url-shortener\\.freecodecamp\\.repl\\.co\\.*')).test(url));
+    }
+    "
 
+  - text: You can pass a URL as a parameter to your app, and you will receive a shortened URL in the JSON response.
+    testString: "async getUserInput => {
+      const url = getUserInput('url');
+      const urlVariable = url.length; 
+      const res = await fetch(url + '/api/shorturl/new/', {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
+        body:  `url=https://fcc-back-end-tester.glitch.me/url-shortener-helper/rand-${urlVariable}`
+      });
+      
+      if (res.ok) {
+        const { short_url, original_url } = await res.json();
+        assert.isNotNull(short_url);
+        assert.match(original_url, new RegExp(`fcc-back-end-tester.glitch.me/url-shortener-helper/rand-${urlVariable}`)); 
+      } else {
+        throw new Error(`${res.status} ${res.statusText}`);
+      }
+    }
+    "
+  
+  - text: When you visit that shortened URL, it will redirect you to your original link.
+    testString: "async getUserInput => {
+      const url = getUserInput('url');
+      const urlVariable = url.length;
+      let shortenedUrlVariable = '';
+      
+      const postResponse = await fetch(url + '/api/shorturl/new/', {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
+        body:  `url=https://fcc-back-end-tester.glitch.me/url-shortener-helper/rand-${urlVariable}`
+      });
+      
+      if (postResponse.ok) {
+        const { short_url } = await postResponse.json();
+        shortenedUrlVariable = short_url
+      } else {
+        throw new Error(`${postResponse.status} ${postResponse.statusText}`);
+      }
+    
+      const getResponse = await fetch(url + '/api/shorturl/' + shortenedUrlVariable)
+      
+      if (getResponse.ok) {
+        const data = await getResponse.json();
+        assert.equal(data.str, `rand-${urlVariable}`); 
+      } else {
+        throw new Error(`${getResponse.status} ${getResponse.statusText}`);
+      }
```

Comment:
Here's the part where I'm not entirely sure if it's best practices or not:

```js
const getResponse = await fetch(url + '/api/shorturl/' + shortenedUrlVariable)
        .then(res => res)

      const getResponse = await fetch(url + '/api/shorturl/' + shortenedUrlVariable)
        .then(res => res)

      if (getResponse) {
        const { redirected, url } = getResponse;

        assert.isTrue(redirected);
        assert.strictEqual(url, `https://url-shortener.freecodecamp.repl.co/rand-${urlVariable}`); 
      } else {
        throw new Error(`${getResponse.status} ${getResponse.statusText}`);
      }
```

My thinking is that, while something like `...projectURL/api/shorturl/1598603478832` would redirect to `https://url-shortener.freecodecamp.repl.co/rand-1598603478832` which 404's, `fetch` still returns enough information for us to determine that the redirect was successful.

Could use more input on this, though. What do you think @RandellDawson and @moT01?